OsmAnd/src/net/osmand/plus/views/GPXLayer.java

```diff
@@ -326,10 +327,57 @@ private void drawSelectedFilesPoints(Canvas canvas, RotatedTileBox tileBox, List
 		for (SelectedGpxFile g : selectedGPXFiles) {
-			List<TrkSegment> points = g.getPointsToDisplay();
-			boolean routePoints = g.isRoutePoints();
-			updatePaints(g.getColor(), routePoints, g.isShowCurrentTrack(), settings, tileBox);
-			drawSegments(canvas, tileBox, points);
+			List<TrkSegment> segments = g.getPointsToDisplay();
+			for (TrkSegment ts : segments) {
+
+				// TODO: Select/install renderables via UI or external process!  (See comments below)
+
+				// Each TrkSegment has one or more 'Renderables' - these handle drawing different things such
+				// as the original track, rainbow altitude colouring, 1km markers, etc. They also handle the
+				// asynchronous re-sampling of tracks for more efficient display. For example, the default
+				// operation Renderable.RenderType.ORIGINAL does Ramer-Douglas-Peucer line optimisation for
+				// zoom-level changes in track resolution. Change the '18' to '20' to see it more clearly.
+
+				// Renderables are processed in order that ALL have asynchronous resampling/culling capabililty.
+				// In an ideal world, the renderers to be used are selected via UI or externally, and the
+				// segment will already have them attached.  For this first version, we add them the very
+				// first time the TrkSegment gets drawn.
+
+				// NOTE: At the moment the 'ORIGINAL' renderer below is TOTALLY FUNCTIONALLY EQUIVALENT								<<<IMPORTANT!!!
+				// with the master branch/version before I added this capability, except that the track now resamples
+				// based on zoom level and displays the resampled version when its available.
+
+				if (ts.renders.size()==0) {  // only do once
+
+					// TODO: To see more clearly the line reduction in action, change the 18 to a higher number (say, 18.5 or 19)...
+
+					ts.addRenderable(view, Renderable.RenderType.ORIGINAL, 18, 0);				// the base line (distance modifier)
+
+					// TODO : enable these to see how the experimental conveyor, altitude, speed, waypoint renders work
+
+					// Note: the conveyor is EXAMPLE ONLY just to show an example of multiple renderables being used
+					// - it is intended to show how support for arrows in route-based rendering can be supported by this
+					// type of system. Please leave the code alone, and I will implement the arrows after this code is approved!
+
+					// You can, of course, comment out the following...
+
+					ts.addRenderable(view, Renderable.RenderType.CONVEYOR, 10, 250);        	// conveyor belt animation (m,refresh(ms))
+					//ts.addRenderable(view, Renderable.RenderType.ALTITUDE, 25, 128);			// an altitude display (m,alpha) << IMPORTANT: See [note 1] below
+					//ts.addRenderable(view, Renderable.RenderType.DISTANCE, 1000, 1);        	// 1km markings (m,size)
+					//ts.addRenderable(view, Renderable.RenderType.SPEED, 20, 255);        		// a speed display (m,alpha) << IMPORTANT: See [Note 1]
+
+					// [Note 1]: The altitude and speed renders (only if enabled, above) have a bug that can crash OsmAnd
```

Comment:
It requires just more testing but it is fine of course would be good to see exception stack trace, I don't think emulator has something very different.